### ISSUE TITLE
Inquiry on private repository authentication for the injector in falcon-system namespace

### ISSUE THREAD

**Issue by RutledgePaulV** [CrowdStrike/falcon-operator#172]
Hello,

I've been using this operator to inject the falcon sensor into pods running in an EKS Fargate cluster. An issue I've run into is that while the injection works fine for images that support public/anonymous access it was encountering an error in the injector webhook running in the falcon-system namespace when it tries to access the image manifest for my private ECR repos since there's no authentication configured. I assume it's fetching the manifest in order to determine the original container command / entrypoint to construct the proxied command / args.

I was able to workaround this issue by configuring a service account with an AWS role that has ECR permissions (IRSA) and then modifying the injector deployment's service account to use the new service account. However, this is not ideal since it requires deploying the operator, waiting for the operator to create the falcon-system namespace with the injector deployment, and then editing the deployment definition to change to service accounts.

I feel as though I missed something in the documentation or that there must be a better way of doing this (as operator config) instead of later modifying the resources the operator creates. Could you steer me in the right direction or otherwise confirm if my workaround is the supported way?

Thank you!

**Comment by isimluk** [comment]
Hello @RutledgePaulV,

All Your observations are correct.

I am now testing a patch set (https://github.com/CrowdStrike/falcon-operator/pull/169) that will teach the operator to collect all the pull-secret tokens on the cluster combine them together into a single one and pass it down to the injector.

I would be curious to learn, whether you have pull-token for your private ECR repo in the kubernetes as a secret. If yes, the patchset #169 should fix this for you.

You can run the following command to get see:
```
kubectl get secrets -n falcon-system-configure | grep kubernetes.io/dockerc
```

**Comment by RutledgePaulV** [comment]
We don't have any kubernetes secrets configured for docker/ECR access aside from the one created by the operator to download images from registry.crowdstrike.com. EKS fargate nodes pull our ECR images using fargate profile permissions and so do not require a kubernetes secret. I'm sure this goes without saying, but I would always prefer to authenticate to services like ECR using IAM roles instead since that way there is no static long-lived token involved. 

```
$ kubectl get secrets -A | grep kubernetes.io/docker
falcon-kubernetes-protection   kpagent-cs-k8s-protection-agent-regsecret        kubernetes.io/dockerconfigjson        1      12d
```

It sounds like once #169 lands that I could create a token in a secret and it would be propagated to the falcon-system namespace by the operator, but could there be a way to support IRSA roles on the injector pod instead?

Thank you for your help.

**Comment by isimluk** [comment]
Nod. That make sense.

Nevertheless, there is something you said that wants me to explore your options in from different perspective.

If your fargate nodes have the access to your private ECR through the AIM. The same injector pod should be able to have the same visibility. Unless of course you happen to separate host network from pod network. In that case pod would not have access to the same IAM role.

In other words, your set-up should just work today.

Note: You may be able to find my e-mail address in the commit history here if you want to continue the conversation more in depth.

**Comment by RutledgePaulV** [comment]
The instance metadata service accessible to the fargate kubelet running on the VM is indeed inaccessible from the pod running on the node (this is done by AWS, nothing I control). In this way pods do not inherit the fargate profile permissions and instead require their own role definitions to access AWS services (like via the service account I setup).

**Comment by isimluk** [comment]
Alright. I wasn't sure about the fargate IAM.

So let me explore one other option.

There are two modes of deployment for Falcon Container described in : https://github.com/CrowdStrike/falcon-operator/tree/main/docs/container#image-registry-considerations 

(Option 2) described there leads into adding AIM role to the operator so the operator can push to your ECR to mirror Falcon Container from CrowdStrike registry to your local ECR. You mentioned in comment 0 here, that you were able to add IAM role to the injector. But it is not clear, whether you also use this mechanism for dealing IAM role to the operator.

If so, we could modify the operator its own permissions to generate pulltoken for the injector. WDYT?

**Comment by RutledgePaulV** [comment]
Yes, I did pursue option 2 to mirror to our own ECR and used a custom service account + iam role for the operator to grant it ECR access. Having the operator generate / maintain its own ECR token for use by the injector would certainly make things easier to bootstrap.

I can think couple decent options:

1. As you suggested - Have the operator maintain a registry token for use by the injector. Cluster administrators only need to ensure the operator has the necessary registry permissions and the operator does the rest. This is easy for administrators but perhaps a little weird to propagate authentication from one pod to another.

2. Provide a way to pass a reference to the operator for a custom service account which should be used by injector when that deployment gets created. This has the benefit of forcing the injector to do its own authentication instead of mooching off of another pod, but then does that mean I also need to create the falcon-system namespace myself since that's where the service account will need to live? Personally I would be fine with that but am unsure how gracefully the operator responds to the namespace already existing.

**Comment by isimluk** [comment]
Thanks this makes lot of sense.

To add to your option 2. Pre-existing namespace is fine. However, the problem is that

- when operator is removed from the cluster it will remove  FalconContainer resource
- when FalconContainer resource is removed from the cluster it will remove anything it had created

Hence, we would have to modify the clean-up code to not remove the service account and its namespace if the service account has OIDC bits assigned to it. (Here I am assuming you used OIDC to deal AIM to SAs).

-- 
So the option 2 is cleaner in terms of handling the secrets. But clean-up and deployment becomes more complex. Which is fine today, but may not be always.

**Comment by RutledgePaulV** [comment]
Personally I would not be surprised to find that when the operator is deleted it also deletes everything in the falcon-system namespace (including a service account I defined) but I understand your position of wanting it to be specific in ownership of resources. Honestly, either approach is something I could live with quite happily.

I'll think about this a bit more and perhaps explore the operator code for inspiration. Thank you for taking the time to understand my problem.

**Comment by isimluk** [comment]
Thank You again for all the input!

I have been exploring the (option 1) to forward operator's pull permissions down to the injector. What I found is that [GetAuthorizationToken](https://docs.aws.amazon.com/AmazonECR/latest/APIReference/API_GetAuthorizationToken.html) expires in 12 hours. Rendering the one-time forward unfeasible.

**Comment by isimluk** [comment]
(Option 3) may be of letting the operator grant the permission to the SA. So, how would that work?

 - user needs to create new iam policy for the image pull
 - user needs to pass policy ARN down to the operator
 - operator policy needs to be changed to be able to grant aim policies on the cluster
 - operator code needs to be amended to deal with the OIDC

Not straight forward either.

**Comment by isimluk** [comment]
Perhaps there should be separate operator that would deal oidc identities to the SAs on the cluster? Perhaps there already is?

**Comment by RutledgePaulV** [comment]
In theory the only thing I need added to the service account is an annotation naming the role I want it to use. It looks like this:

```
apiVersion: v1
automountServiceAccountToken: true
kind: ServiceAccount
metadata:
  annotations:
    eks.amazonaws.com/role-arn: arn:aws:iam::<my-account-id>:role/falcon-injector-k8s-svc
  creationTimestamp: "2022-02-01T19:07:05Z"
  name: falcon-injector
  namespace: falcon-system
  resourceVersion: "7130199"
  uid: e01d5a37-9481-47c4-88b8-4cf688ba8a97
secrets:
- name: falcon-injector-token-mv2kc
```

If the operator could just propagate a map of annotations onto a non-default service account in falcon-system namespace I think that would cover it.
